### PR TITLE
Initial dict_factory support for asdict.

### PR DESCRIPTION
**Pull request by Tinche**
No docs yet, will come later.

As you see, the change in non-test code is minimal. 

In the test code, I've tried integrating Hypothesis more. We can now generate classes with attributes using Hypothesis. We generate only attributes with defaults, so the classes can be instantiated without prior knowledge of their attributes, and no subclasses.

### CODE REVIEW

**Comment by hynek** [comment]
I’m not sure when I’ll have idle cycles to review proper, but I’m a big fan of more hypothesis. :)

(if you get bored, please make docstrings `"""\ndoc\n"""` :))

File: tests/test_funcs.py [hunk 1]
```diff
@@ -66,14 +75,41 @@ def test_lists_tuples(self, container, C):
...
             "x": 1,
             "y": {"a": {"x": 4, "y": 5}},
-        } == asdict(C(1, {"a": C(4, 5)}))
+        } == res
+        assert isinstance(res, dict_factory)
+
+    @given(simple_classes, st.sampled_from([dict, OrderedDict]))
+    def test_roundtrip(self, cls, dict_factory):
+        """
+        Test roundtripping for Hypothesis-generated classes.
```
Comment (hynek) [inline tests/test_funcs.py]: maybe explain in a few words, what roundtripping means :)

File: tests/test_funcs.py [hunk 2]
```diff
@@ -3,13 +3,19 @@
 """
 
 from __future__ import absolute_import, division, print_function
+from collections import OrderedDict
```
Comment (hynek) [inline tests/test_funcs.py]: empty line after __future__ please

File: tests/test_funcs.py [hunk 3]
```diff
@@ -3,13 +3,19 @@
...
 
+from hypothesis import given, strategies as st
+
+from . import simple_classes
+
 from attr._funcs import (
     asdict,
     assoc,
     has,
+    fields,
```
Comment (hynek) [inline tests/test_funcs.py]: i usually try to sort my imports if they’re one/line :)

File: tests/__init__.py [hunk 4]
```diff
@@ -40,3 +44,20 @@ def returns_distinct_classes(self):
         Each call returns a completely new class.
         """
         assert simple_class() is not simple_class()
+
+
+def create_class(attrs):
+    """
+    A helper function for Hypothesis to generate attrs classes.
+    """
+    # What if we get more than len(string.ascii_lowercase) attributes?
```
Comment (hynek) [inline tests/__init__.py]: you tell me :)

File: src/attr/_funcs.py [hunk 5]
```diff
@@ -22,10 +22,15 @@ def asdict(inst, recurse=True, filter=None):
         value as the second argument.
     :type filer: callable
 
+    :param dict_factory: A callable to produce dictionaries from. For example,
+        to produce ordered dictionaries instead of normal Python dictionaries,
+        pass in ``collections.OrderedDict``.
+    :type dict_factory: callable
```
Comment (hynek) [inline src/attr/_funcs.py]: this should be written as `:param callable dict_factory:`.  Bonus points if you fix it above too. :)

File: tests/__init__.py [hunk 6]
```diff
@@ -1,6 +1,10 @@
 from __future__ import absolute_import, division, print_function
 
-from attr import Attribute
+import string
+
+from hypothesis import strategies as st
+
+from attr import Attribute, ib
```
Comment (hynek) [inline tests/__init__.py]: I find using `ib` on itself weird and confusing.

Just import additionally `attr` and use `attr.ib`.

File: tests/__init__.py [hunk 7]
```diff
@@ -40,3 +44,20 @@ def returns_distinct_classes(self):
         Each call returns a completely new class.
         """
         assert simple_class() is not simple_class()
+
+
+def create_class(attrs):
```
Comment (hynek) [inline tests/__init__.py]: The naming here is a bit confusing given `simple_class` above.  Maybe `hyp_create_class`?  Or make it private

File: tests/test_funcs.py [hunk 8]
```diff
@@ -21,16 +27,18 @@ class TestAsDict(object):
     """
     Tests for `asdict`.
     """
-    def test_shallow(self, C):
+    @given(st.sampled_from([dict, OrderedDict]))
```
Comment (hynek) [inline tests/test_funcs.py]: Let’s have a `DICT_TYPES = (dict, OrderedDict,)`.  Maybe we’ll get more one day and this is quite repetitive.

**Comment by hynek** [comment]
Additionally:

- `dict_factory` needs a `.. versionadded::` tag
- CHANGELOG

Thanks. :)

**Comment by Tinche** [comment]
Roger, thanks for the comments. I'll get around to it in a few days.

**Comment by Tinche** [comment]
Ping. :)

**Comment by hynek** [comment]
Seems like you broke flake8 & docs. :)

File: CHANGELOG.rst [hunk 9]
```diff
@@ -28,6 +28,9 @@ Changes:
 - ``__slots__`` have arrived!
   Classes now can automatically be `slots <https://docs.python.org/3.5/reference/datamodel.html#slots>`_-style (and save your precious memory) just by passing ``slots=True``.
   `#35 <https://github.com/hynek/attrs/issues/35>`_
+- ``attr.asdict`` can now produce arbitrary mappings instead of Python ``dict``s
+  when provided a ``dict_factory`` argument.
```
Comment (hynek) [inline CHANGELOG.rst]: “…when provided *with* a…”

Also please merge those two lines. just keep the link in a separate line.

**Comment by Tinche** [comment]
Hm, docs are still broken but for an unrelated reason:

`Invalid requirement: 'zope.interface  # for doctests'`

Maybe pip on Travis can't handle the comment? (7.1.2) `tox -e docs` works just fine for me locally.

**Comment by hynek** [comment]
could you try removing it then?

**Comment by Tinche** [comment]
That seems to have fixed it. How did it work before?

**Comment by hynek** [comment]
`¯\_(ツ)_/¯`

Thanks!

**Comment by Tinche** [comment]
\o/

**Pull request merged**